_posts/2024-01-25-websockets-in-http4s.md

```diff
@@ -0,0 +1,1561 @@
+
+Create `Util.scala` in the following path, `src/main/scala/rockthejvm/websockets/Util.scala` and add the following code:
+
+```scala
+package rockthejvm.websockets
+
+import cats.data.Validated
+import cats.syntax.all.*
+
+def validateItem[F](
+    value: String,
+    userORRoom: F,
+    name: String
+): Validated[String, F] =
+  Validated.cond(
+    (value.length >= 2 && value.length <= 10),
+    userORRoom,
+    s"$name must be between 2 and 10 characters"
+  )
+
+case class User(name: String)
+object User {
+  def apply(name: String): Validated[String, User] =
+    validateItem(name, new User(name), "User name")
+}
+
+case class Room(room: String)
+object Room{
+  def apply(room: String): Validated[String, Room] =
+    validateItem(room, new Room(room), "Room")
+}
+
+case class ChatState(
+    userRooms: Map[User, Room],
+    roomMembers: Map[Room, Set[User]]
+)
+object ChatState:
+  def apply(userRooms: Map[User, Room], roomMembers: Map[Room, Set[User]]) =
+    new ChatState(userRooms, roomMembers)
+```
+
+Here we create a `User` and `Room` case classes both of which take a value of type `String` and within each companion object, we call `validateItem()` in the `apply()` method to validate their inputs. 
+
+The `validateItem()` function takes as parameters a `String` input from the `apply()` function, an instance of the case class, and a `name` of type `String` that becomes part of an error message. 
+
+In the body, we use `Validated.cond()` function to produce a `User` or `Room` if the value is between 2 and 10 characters since we don't want very short or very long user/room names.
+
+Lastly, we have `ChatState`, a case class that holds all the rooms and users in the chat application at a particular moment. 
+
+It takes two parameters, `userRooms` of type `Map[User, Room]` that shows which room each user is in, and `roomMembers` of type `Map[Room, Set[User]]` that shows all the users in each room.
+
+### 6.2. The Output Message
+Next, we create an `OutputMessage` ADT to represent the messages sent to the WebSocket client. Create `OutputMessage.scala` and add the following code:
+
+```scala
+package rockthejvm.websockets
+
+sealed trait OutputMessage
+
+case class Register(
+    user: Option[User],
+    msg: String = """
+        |Register your username with the following command:
+        |/name <username>
+        """.stripMargin
+) extends OutputMessage
+
+case class ParsingError(user: Option[User], msg: String) extends OutputMessage
+```
+
+Here we create our first two `OutputMessage` messages, `Register` and `ParsingError`, that both take a `user` of type `Option[User]` and a `msg` of type `String` which instructs the user how to register or passes an error message in case registration failed respectively:
+
+```scala
+sealed trait OutputMessage
+...
+case class SuccessfulRegistration(
+    user: User,
+    msg: String = ""
+) extends OutputMessage
+object SuccessfulRegistration {
+  def apply(user: User) = {
+    user match
```

Comment:
braces here

I don't want to cover the entire review with just "use braces" comments, I trust you'll make the changes on all the code snippets

Reply:
Fixed braces throughout the code